pkg/controller/machinedeployment/machinedeployment_controller_test.go

```diff
@@ -94,9 +100,58 @@ func TestReconcile(t *testing.T) {
 	g.Expect(c.Delete(context.TODO(), &ms)).NotTo(gomega.HaveOccurred())
 	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
 	g.Eventually(errors, timeout).Should(gomega.Receive(gomega.BeNil()))
+	g.Eventually(func() int {
+		if err := c.List(context.TODO(), &client.ListOptions{}, machineSets); err != nil {
+			return -1
+		}
+		return len(machineSets.Items)
+	}, timeout).Should(gomega.BeEquivalentTo(1))
+
+	// Scale a MachineDeployment and expect Reconcile to be called
+	err = updateMachineDeployment(c, instance, func(d *clusterv1alpha1.MachineDeployment) { d.Spec.Replicas = int32Ptr(5) })
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Expect(c.Update(context.TODO(), instance)).NotTo(gomega.HaveOccurred())
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+	g.Eventually(errors, timeout).Should(gomega.Receive(gomega.BeNil()))
+	g.Eventually(func() int32 {
+		if err := c.List(context.TODO(), &client.ListOptions{}, machineSets); err != nil {
+			return -1
+		}
+		if len(machineSets.Items) != 1 {
+			return -1
+		}
+		return *machineSets.Items[0].Spec.Replicas
+	}, timeout).Should(gomega.BeEquivalentTo(5))
+
+	// Update a MachineDeployment, expect Reconsile to be called and a new MachineSet to appear
+	err = updateMachineDeployment(c, instance, func(d *clusterv1alpha1.MachineDeployment) { d.Spec.Template.Labels["updated"] = "true" })
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	g.Expect(c.Update(context.TODO(), instance)).NotTo(gomega.HaveOccurred())
+	g.Eventually(requests, timeout).Should(gomega.Receive(gomega.Equal(expectedRequest)))
+	g.Eventually(errors, timeout).Should(gomega.Receive(gomega.BeNil()))
+	g.Eventually(func() int {
+		if err := c.List(context.TODO(), &client.ListOptions{}, machineSets); err != nil {
+			return -1
+		}
+		return len(machineSets.Items)
+	}, timeout).Should(gomega.BeEquivalentTo(2))
```

Comment:
Does the update finish by removing the original machine set?

Reply:
No, removing old `MachineSets` does not work at all (and never did). Getting that fixed was the actual issue I wanted to resolve. When writing a test for that and rebasing it into the crd-based version of the `MachineDeployment` controller I realized there are currently no tests at all, which is why I created these PRs :)

Reply:
Can you file an issue to describe what doesn't work?

Reply:
#543